Optimize the implementation of the argsort operator.

Vvsmile opened the pull request:
<!-- Demo: https://github.com/PaddlePaddle/Paddle/pull/24810 -->
### PR types
<!-- One of [ New features | Bug fixes | Function optimization | Performance optimization | Breaking changes | Others ] -->
Performance optimization
### PR changes
<!-- One of [ OPs | APIs | Docs | Others ] -->
OPs
### Describe
<!-- Describe what this PR does -->
Optimize the implementation of argsort operator. The performance is better than Pytorch.  
Given on the input size of [4, 68992], which is slower than Pytorch initially, the performance of optimized argsort kernel can achieve 3.3x speedup, which is ~1.5x the performance of Pytorch. 
Reorganize the code of argsort operator to reduce the code redundancy.  
This optimal codes are called when the number of rows of input tensor is less.  
When the number of input tensor is large, the original implementation is called. (The performance is also better than Pytorch)  

| case | pytorch | paddle (before optimization) | diff | paddle (after optimization) | diff | speedup |
|------|--------|-----|------|------|------|-------|
|x(Variable)-dtype: float32, shape: [1L, 68892L]  axis(int):-1 | 0.17 | 0.23 | lower than(35.3%) | 0.20| lower than(17.6%)| 1.15|
|x(Variable)-dtype: float32, shape: [2L, 68892L]  axis(int):-1|0.25|0.91|lower than(2.64)|0.28|lower than(12%)|3.25|
|x(Variable)-dtype: float32, shape: [4L, 68892L]  axis(int):-1|0.42|0.97|lower than(1.31)|0.29|greater than(31.0%)|3.34|
|x(Variable)-dtype: float32, shape: [8L, 68892L]  axis(int):-1|0.71|1.03|lower than(45.1%)|0.39|greater than(45.1%)|2.64|
|x(Variable)-dtype: float32, shape: [16L, 68892L]  axis(int):-1|1.21|1.07|greater than(11.6%)|1.04|greater than(14.0%)|1.03|
|x(Variable)-dtype: float32, shape: [32L, 68892L]  axis(int):-1|2.23|1.31|greater than(41.3%)|1.11|greater than(50.2%)|1.18|

Vvsmile committed:
Optimize the implementation of the argsort operator.
Reorganize the code to reduce the code redundancy.

The optimal codes are called when the number of rows is limited
in the input tensor.

modification:
	paddle/phi/kernels/gpu/argsort_kernel.cu

test=develop

Vvsmile committed:
Optimize the implementation of the argsort operator.
Reorganize the code to reduce the code redundancy.

The optimal codes are called when the number of rows is limited
in the input tensor.

modification:
	paddle/phi/kernels/gpu/argsort_kernel.cu

test=develop

ZzSean commented on paddle/phi/kernels/gpu/argsort_kernel.cu:
```diff
@@ -76,13 +76,285 @@ static __global__ void FillIndex(T* indices, T num_rows, T num_cols) {
 }
 
+// this is defined for CUDA LOOP
+#define CUDA_KERNEL_LOOP_TYPE(i, n, index_type)               \
```
这个宏在Paddle中已经定义好了，在Paddle/paddle/phi/backends/gpu/cuda/cuda_helper.h

Vvsmile replied:
好的,那我把这里给去掉

ZzSean commented on paddle/phi/kernels/gpu/argsort_kernel.cu:
```diff
@@ -76,13 +76,285 @@ static __global__ void FillIndex(T* indices, T num_rows, T num_cols) {
+       _i_n_d_e_x += blockDim.x * gridDim.x, i = _i_n_d_e_x)
+#define CUDA_KERNEL_LOOP(i, n) CUDA_KERNEL_LOOP_TYPE(i, n, int)
+
+static __global__ void Fill_index_and_segment_kernel(int2 *data,
```
1. 这里为什么要加static
2. 函数名命名规则用大驼峰

Vvsmile replied:
我换成大驼峰去,
外面没有这里的调用,想着直接定在模块里面,也可以去掉,全局里没有同名函数

ZzSean commented on paddle/phi/kernels/gpu/argsort_kernel.cu:
```diff
@@ -76,13 +76,285 @@ static __global__ void FillIndex(T* indices, T num_rows, T num_cols) {
+
+constexpr int CUDA_NUM_THREADS = 1024;
+
+inline int GET_BLOCKS(const int64_t N,
```
同上

Vvsmile replied:
okk

ZzSean commented on paddle/phi/kernels/gpu/argsort_kernel.cu:
```diff
@@ -76,13 +76,285 @@ static __global__ void FillIndex(T* indices, T num_rows, T num_cols) {
+
+inline int GET_BLOCKS(const int64_t N,
+                      const int64_t max_threads_per_block = CUDA_NUM_THREADS) {
+  constexpr int64_t max_int = std::numeric_limits<int>::max();
```
这个值好像没用到

Vvsmile replied:
存在这个冗余,已删除

ZzSean commented on paddle/phi/kernels/gpu/argsort_kernel.cu:
```diff
@@ -277,21 +509,29 @@ void ArgsortKernel(const Context& dev_ctx,
+    }
 
     TransposeKernel<int64_t, Context>(dev_ctx, tmp_indices, trans, indices);
-    // transpose back
```
这些注释不用删吧

Vvsmile replied:
我把这个注释换上,当初在这些里面读代码的时候加了注释,后面删删多了

ZzSean commented on paddle/phi/kernels/gpu/argsort_kernel.cu:
```diff
@@ -133,96 +405,49 @@ void ArgFullSort(const phi::GPUContext& ctx,
+  int64_t *ids_data = dev_ctx.template Alloc<int64_t>(indices);
+  const T *in_data = input.data<T>();
 
-  // Use thrust for parallel acceleration when the input size is equal to the
```
这些注释为什么要删掉呢

Vvsmile replied:
同上

ZzSean commented on paddle/phi/kernels/gpu/argsort_kernel.cu:
```diff
@@ -133,96 +405,49 @@ void ArgFullSort(const phi::GPUContext& ctx,
-                   DenseTensor* output,
-                   DenseTensor* indices) {
-  auto in_dims = input.dims();
-  axis = (axis < 0) ? (in_dims.size() + axis) : axis;
```
这句也不能删吧？

Vvsmile replied:
嗯嗯,确实不能

ZzSean commented on paddle/phi/kernels/gpu/argsort_kernel.cu:
```diff
@@ -76,13 +76,285 @@ static __global__ void FillIndex(T* indices, T num_rows, T num_cols) {
+      self_ptr, values_ptr, indices_ptr, i_s_ptr, nsegments, nsort);
+}
+
+// Sort by flag descending, True: descending. False: Ascending.
```
注释格式有点乱

Vvsmile replied:
我全都过一遍

ZzSean commented on paddle/phi/kernels/gpu/argsort_kernel.cu:
```diff
@@ -76,13 +76,285 @@ static __global__ void FillIndex(T* indices, T num_rows, T num_cols) {
+                            const IndType num_rows,
+                            const IndType num_cols,
+                            const bool descending) {
+  auto cu_stream = ctx.stream();
```
`cu_stream`->`gpu_stream`

Vvsmile replied:
好的

ZzSean commented on paddle/phi/kernels/gpu/argsort_kernel.cu:
```diff
@@ -76,13 +76,285 @@ static __global__ void FillIndex(T* indices, T num_rows, T num_cols) {
+  ctx.template Alloc<IndType>(&input_indices);
+  size_t temp_storage_bytes = -1;
+
+  int block_size = 1024;
```
不用魔鬼数字，用你前面宏定义的`CUDA_NUM_THREADS`

Vvsmile replied:
这个我改掉了,还没有推上来

ZzSean commented on paddle/phi/kernels/gpu/argsort_kernel.cu:
```diff
@@ -76,13 +76,285 @@ static __global__ void FillIndex(T* indices, T num_rows, T num_cols) {
+
+  int block_size = 1024;
+  int maxGridDimX = ctx.GetCUDAMaxGridDimSize()[0];
+  int grid_size = num_rows < maxGridDimX ? num_rows : maxGridDimX;
```
你这里的grid和block设置跟最上面的`GET_BLOCKS`两者之间好像没有什么关系？尽量把相同相似的逻辑抽取出来，封装使用，避免重复代码
而且感觉你这里算的block和grid并没有用到啊

Vvsmile replied:
这个我已经删掉了,还没有推上来

Vvsmile committed:
Optimize the Argsort operator and merge the remote branch 'develop'.

This version passes all of the unittests.

Merge action:
	Merge branch 'develop' of https://github.com/Vvsmile/Paddle into develop

Vvsmile committed:
Optimize the Argsort operator and merge the remote branch 'develop'.

This version passes all of the unittests.

Merge action:
	Merge branch 'develop' of https://github.com/Vvsmile/Paddle into develop

Vvsmile committed:
Merge branch 'develop' of https://github.com/Vvsmile/Paddle into develop

Modify the argsort operator to pass the speed testing.

ZzSean commented on paddle/phi/kernels/gpu/argsort_kernel.cu:
```diff
@@ -64,8 +64,12 @@ struct SegmentOffsetIter {
   int num_cols_;
 };
 
+// using phi::paddle::platform::PADDLE_CUDA_NUM_THREADS;
```
这个可以删

Vvsmile replied:
好嘞

ZzSean commented on paddle/phi/kernels/gpu/argsort_kernel.cu:
```diff
@@ -76,13 +80,270 @@ static __global__ void FillIndex(T* indices, T num_rows, T num_cols) {
+    PADDLE_ENFORCE_GPU_SUCCESS(err);                                           \
+  } while (false)
+
+inline int GET_BLOCKS(
```
函数名大驼峰，另外如果这个函数功能这么简单的话，是不是可以不用封装了

Vvsmile replied:
明白,我把这个弄一下

ZzSean commented on paddle/phi/kernels/gpu/argsort_kernel.cu:
```diff
@@ -115,114 +376,68 @@ void ArgFullSort(const phi::GPUContext& ctx,
+  auto size = input.numel();
   auto in_dims = input.dims();
   axis = (axis < 0) ? (in_dims.size() + axis) : axis;
-  const T* in_data = input.data<T>();
```
这部分感觉不用改

Vvsmile replied:
好嘞

Vvsmile replied:
这个我查了一下,没有删掉,就是我之前编辑的时候确实不小心删掉过,但是我又加上了,只是位置不在这个地方,在这个后两行的位置,不影响程序执行

Vvsmile committed:
Modify the codes based on the suggestions of PR review.

Delete some unuseful codes and change the function name to hump format.

ZzSean commented on paddle/phi/kernels/gpu/argsort_kernel.cu:
```diff
@@ -76,13 +78,261 @@ static __global__ void FillIndex(T* indices, T num_rows, T num_cols) {
+  }
+}
+
+template <typename scalar_t>
```
模版参数还是用T吧，下同

ZzSean commented on paddle/phi/kernels/gpu/argsort_kernel.cu:
```diff
@@ -76,13 +78,261 @@ static __global__ void FillIndex(T* indices, T num_rows, T num_cols) {
+                         segment_bits,
+                         ctx);
+
+  SortPostprocessKernel<<<(n + 511) / 512, 512, 0, cu_stream>>>(
```
这个block和grid写法有点暴力啊。。还是用变量吧，或者上面的结果不能直接用吗

Vvsmile replied:
换成了上面的结果形式

ZzSean commented on paddle/phi/kernels/gpu/argsort_kernel.cu:
```diff
@@ -76,13 +78,261 @@ static __global__ void FillIndex(T* indices, T num_rows, T num_cols) {
+
+// The method is called when # of the rows of the input is less than or equal to
+// 32
+template <typename T, typename IndType>
```
`IndType`->`IndexType`

Vvsmile replied:
已修

Vvsmile committed:
Update argsort_kernel.cu

Vvsmile committed:
fix the difference from original file

the difference should not be changed

Vvsmile committed:
Fix some codes based on the suggestions

ZzSean commented:
优化效果在PR描述中补充一下

Vvsmile committed:
fix code style of argsort_kernel.cu

ZzSean commented on paddle/phi/kernels/gpu/argsort_kernel.cu:
```diff
@@ -78,23 +81,273 @@ static __global__ void FillIndex(T* indices, T num_rows, T num_cols) {
+template <typename KT, typename VT>
+static void RadixSortPairsImpl(const KT *keys_in,
+                               KT *keys_out,
+                               const VT *values_in,
```
变量顺序把输入和输出放在一起，输入在前输出在后

Vvsmile replied:
好的

ZzSean commented on paddle/phi/kernels/gpu/argsort_kernel.cu:
```diff
@@ -78,23 +81,273 @@ static __global__ void FillIndex(T* indices, T num_rows, T num_cols) {
+                               bool descending,
+                               int64_t begin_bit,
+                               int64_t end_bit,
+                               const phi::GPUContext &ctx) {
```
这个`ctx`参数放在最前面

ZzSean commented on paddle/phi/kernels/gpu/argsort_kernel.cu:
```diff
@@ -78,23 +81,273 @@ static __global__ void FillIndex(T* indices, T num_rows, T num_cols) {
+                               const phi::GPUContext &ctx) {
+  if (keys_out == nullptr) {
+    DenseTensor key_out_owner;
+    int64_t key_out_owner_size = n;
```
直接用n不就可以

ZzSean commented on paddle/phi/kernels/gpu/argsort_kernel.cu:
```diff
@@ -78,23 +81,273 @@ static __global__ void FillIndex(T* indices, T num_rows, T num_cols) {
+template <typename KT, typename VT>
+static void RadixSortPairs(const phi::GPUContext &ctx,
+                           const KT *keys_in,
+                           KT *keys_out,
```
同上

ZzSean commented on paddle/phi/kernels/gpu/argsort_kernel.cu:
```diff
@@ -78,23 +81,273 @@ static __global__ void FillIndex(T* indices, T num_rows, T num_cols) {
+}
+
+template <typename KT, typename VT>
+static void RadixSortPairs(const phi::GPUContext &ctx,
```
这个函数跟`RadixSortPairsImpl`有什么区别吗？好像没看到什么别的处理

ZzSean commented on paddle/phi/kernels/gpu/argsort_kernel.cu:
```diff
@@ -78,23 +81,273 @@ static __global__ void FillIndex(T* indices, T num_rows, T num_cols) {
+
+template <typename KT>
+static void RadixSortKeys(const KT *keys_in,
+                          KT *keys_out,
```
同上

ZzSean commented on paddle/phi/kernels/gpu/argsort_kernel.cu:
```diff
@@ -78,23 +81,273 @@ static __global__ void FillIndex(T* indices, T num_rows, T num_cols) {
+
+  T *sorted_out_ptr;
+  IndexType *sorted_indices_ptr;
+  const T *inp = input->data<T>();
```
这个命名也改下`inp`->`input_data`
下同

ZzSean commented on paddle/phi/kernels/gpu/argsort_kernel.cu:
```diff
@@ -78,23 +81,273 @@ static __global__ void FillIndex(T* indices, T num_rows, T num_cols) {
+                                         const int64_t nsort,
+                                         const int64_t n,
+                                         const bool descending,
+                                         const T *const self_ptr,
```
参数位置调整，一般是：context，输入，输出，其他参数
其他函数有类似问题的也改下

ZzSean commented on paddle/phi/kernels/gpu/argsort_kernel.cu:
```diff
@@ -1,3 +1,4 @@
+
```
空行删掉

ZzSean commented on paddle/phi/kernels/gpu/argsort_kernel.cu:
```diff
@@ -78,23 +81,273 @@ static __global__ void FillIndex(T* indices, T num_rows, T num_cols) {
+                            const IndexType num_cols,
+                            const bool descending) {
+  auto gpu_stream = ctx.stream();
+  DenseTensor input_indices;
```
这个tensor在你的函数里没用到啊

Vvsmile committed:
fix argsort based on suggestions

Vvsmile committed:
fix code style of argsort_kernel

Vvsmile committed:
fix code style of argsort(gpu version)

Vvsmile committed:
fix the no instance of function template SegmentedSoftPairsByFullSort
match the argument list

Vvsmile committed:
remove empty line of argsort_kernel

Vvsmile committed:
fix the order of arguments list of function SortPostprocessKernel

Vvsmile committed:
fix arguments of SortPostprocessKernel

Vvsmile committed:
change the test_argsort_op to enable_static

Vvsmile committed:
fix the file test_argsort_op.py

ZzSean reviewed (approved):
LGTM
